File: solutions/go/0090_子集 II.go
package main

import "sort"

// 90. 子集 II
// 排序 + 回溯：同一层若 nums[i] == nums[i-1] 则跳过，避免重复子集。
func subsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	result := [][]int{}
	path := []int{}

	var backtrack func(start int)
	backtrack = func(start int) {
		// 必须拷贝 path
		cur := make([]int, len(path))
		copy(cur, path)
		result = append(result, cur)
		for i := start; i < len(nums); i++ {
			if i > start && nums[i] == nums[i-1] {
				continue
			}
			path = append(path, nums[i])
			backtrack(i + 1)
			path = path[:len(path)-1]
		}
	}
	backtrack(0)
	return result
}

File: solutions/java/0090_子集 II.java
public class Solution {
    /**
     * 90. 子集 II
     * 排序 + 回溯：同一层若 nums[i] == nums[i-1] 则跳过，避免重复子集。
     */
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        java.util.Arrays.sort(nums);
        List<List<Integer>> result = new java.util.ArrayList<>();
        List<Integer> path = new java.util.ArrayList<>();
        backtrack(nums, 0, path, result);
        return result;
    }

    private void backtrack(int[] nums, int start, List<Integer> path, List<List<Integer>> result) {
        result.add(new java.util.ArrayList<>(path));
        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1]) {
                continue;
            }
            path.add(nums[i]);
            backtrack(nums, i + 1, path, result);
            path.remove(path.size() - 1);
        }
    }
}